### PR TITLE
Improve C++ build

### PR DESCRIPTION
**Pull request by gspowley**
Changes to address issues found building the C++ library on different hosts:
* Update build from source docs and `./scripts/setup` script to remove old build files.
* Fix runtime error: `ImportError: generic_type: type "PyQueryCondition" is already registered!`
* Update `./scripts/bld` script to support building on ARM.

### CODE REVIEW

**Comment by johnkerl** [comment]
@gspowley re

> One more issue to solve:
> `ImportError: dynamic module does not define module export function (PyInit_libtiledbsoma)`

Do we have a handle on under which circumstances this does (and does not) happen?

**Comment by gspowley** [comment]
Yes. On some systems, the "real" `libtiledbsoma.cpython-310-aarch64-linux-gnu.so` is built properly, but `pip` install builds an "empty" `libtiledbsoma.cpython-310-aarch64-linux-gnu.so`, which replaces the real one.

Some evidence from `pip install -v -e .`:
```
...
  Built target install-libtiledbsoma
    copying file /home/ubuntu/work/TileDB-SOMA/apis/python/../../dist/lib/libtiledbsoma.cpython-310-aarch64-linux-gnu.so to /home/ubuntu/work/TileDB-SOMA/apis/python/src/tiledbsoma
    copying file /home/ubuntu/work/TileDB-SOMA/apis/python/../../dist/lib/libtiledb.so.2.11 to /home/ubuntu/work/TileDB-SOMA/apis/python/src/tiledbsoma
    adding to package_data: ['libtiledbsoma.cpython-310-aarch64-linux-gnu.so', 'libtiledb.so.2.11']
  building 'tiledbsoma.libtiledbsoma' extension
  creating /tmp/tmp74_bclef.build-lib/tiledbsoma
  aarch64-linux-gnu-gcc -shared -Wl,-O1 -Wl,-Bsymbolic-functions -Wl,-Bsymbolic-functions -g -fwrapv -O2 -L/usr/lib/aarch64-linux-gnu -o /tmp/tmp74_bclef.build-lib/tiledbsoma/libtiledbsoma.cpython-310-aarch64-linux-gnu.so
  copying /tmp/tmp74_bclef.build-lib/tiledbsoma/libtiledbsoma.cpython-310-aarch64-linux-gnu.so -> src/tiledbsoma
...
```

```
-rwxrwxr-x 1 ubuntu ubuntu   7680 Sep 19 16:30 /home/ubuntu/work/TileDB-SOMA/apis/python/src/tiledbsoma/libtiledbsoma.cpython-310-aarch64-linux-gnu.so
-rw-r--r-- 1 ubuntu ubuntu 826104 Sep 19 16:30 /home/ubuntu/work/TileDB-SOMA/dist/lib/libtiledbsoma.cpython-310-aarch64-linux-gnu.so
```

**Comment by johnkerl** [comment]
Sorry, I approved this while it was still in draft (all my excitement!!) -- sorry to not have respected your wishes 🙏

**Comment by johnkerl** [comment]
2nd question while I'm here -- do we have tracking for MacOS build failures or should I open a story?

```
...
      gcc -bundle -undefined dynamic_lookup -arch x86_64 -g -o /var/folders/7l/_wsjyk5d4p3dz3kbz7wxn7t00000gn/T/tmpz4oaepn9.build-lib/tiledbsoma/libtiledbsoma.cpython-39-darwin.so
      clang: error: no input files
...
```

https://gist.github.com/johnkerl/b60fbb291571816c304fe66d111e356e

**Comment by gspowley** [comment]
No problem, I thought the issue was fixed, but it came up again with more testing.

I believe the macos `clang: error: no input files` is related to the "building" an empty shared object issue.

**Comment by johnkerl** [comment]
> @johnkerl please try to build on macos with this branch. We can open a new story if macos still has build issues.

@gspowley I just did an hour or so ago before writing my previous comment but will try again now

**Comment by johnkerl** [comment]
@gspowley https://github.com/single-cell-data/TileDB-SOMA/pull/304/commits/9e2d28e2f33e49c19957da47a516c8177f8d6c80 is a big step forward!!! 🥳 Now I do get a successful build on MacOS, although an error at import time.

https://gist.github.com/johnkerl/b7edcad764a8c46619cb1d7f47a4add5

File: scripts/setup [hunk 1]
```diff
@@ -9,14 +9,19 @@ cd $(git rev-parse --show-toplevel)
 python -m venv test/tiledbsoma
 source test/tiledbsoma/bin/activate
 
-# install python api and dependencies
+# clean
+rm -rf build dist
+rm -f apis/python/src/tiledbsoma/libtiledb.so.*
```
Comment (johnkerl) [inline scripts/setup]: note these end in `.dylib` on MacOS

**Comment by gspowley** [comment]
Thanks @johnkerl!! This looks similar to the pre-installed `libfmt` issues we saw on Linux.

I'll merge these changes and we can deal with `libfmt` separately.

File: scripts/setup (hunk 1 above)
Comment (gspowley) [reply scripts/setup]: Right, that was updated in the [29e230a](https://github.com/single-cell-data/TileDB-SOMA/pull/304/commits/29e230a7ca74aefe9d2b72e55e88fcb84efb8453). Maybe the review picked up an old version?

**Pull request merged**